File: koinKembalian.cpp
#include <bits/stdc++.h>

using namespace std;
int n, c;
int coins[500], memo[50000];
bool computed[50000];

int solve(int x)
{
    if (x == 0)
        return 0;
    if (computed[x])
        return memo[x];

    int best = 5000;

    for (int i = 0; i < n; i++)
    {
        if (coins[i] <= x)
        {
            best = min(best, solve(x - coins[i]) + 1);
        }
    }

    computed[x] = true;
    memo[x] = best;

    return best;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    cin >> n;
    for (int i = 0; i < n; i++)
        cin >> coins[i];
    // get input of coins
    cin >> c;
    cout << solve(c) << endl;
    return 0;
}

File: koinKembalian.py
def amount_of_change_coins(change, coins, computed, memo):
    if change == 0:
        return 0

    if computed[change]:
        return memo[change]

    best = 50000
    for coin in coins:
        if coin <= change:
            best = min(best, amount_of_change_coins(
                change-coin, coins, computed, memo)+1)

    computed[change] = True
    memo[change] = best

    return best


if __name__ == "__main__":
    n = int(input())
    coins = list(map(int, input().split()))
    change = int(input())

    computed = []
    for _ in range(change+1):
        computed.append(False)

    memo = []
    for _ in range(change+1):
        memo.append(0)

    result = amount_of_change_coins(change, coins, computed, memo)
    if result == 50000:
        print(-1)
    else:
        print(result)
